go/chat/crypto_doc.md

```diff
@@ -0,0 +1,205 @@
+  later one of her devices will check @bobbymcferrin's proof and send him keys
+  automatically.
+- If either of them adds a new device, that device will get a copy of all the
+  keys it needs to read old messages.
+- If either of them removes a device, all of their other devices will start
+  using a new encryption key. That guarantees the removed device can't read new
+  messages, without relying on the server to enforce it.
+
+Even though the server can't forge new messages, there are tricks it might try
+to play. For performance reasons, it's the server's responsibility to assign a
+sequential ID to each message. An evil server could try to reorder some
+messages, for example, or leave some of them out. To limit tricks like that,
+every time Alice sends a message, she includes some references to other
+messages she's seen before. Bob's device will check that list, to make sure
+it's consistent with what he's seeing. That limits the mischief an evil server
+can do, while still letting Alice send messages quickly on a crappy network.
+
+That shared context prevents the server from dropping Alice's messages without
+her permission, but it's still important that Alice can choose to delete them
+when she wants to. To make that work without losing the context, every message
+comes in two parts, a header and a body. The header holds the previous message
+references and never gets deleted. The body holds the text of a message, and
+when Alice sends a special "delete" message, the server will delete it. That
+way Bob can check that the deletion is allowed. Editing messages works in a
+similar way; Alice sends a special "edit" message that Bob can verify.
+
+Signing and encrypting attachments is almost the same as for regular messages,
+except that they could be very large. If Alice uploads a video, Bob might want
+to skip to some little part of it without downloading the whole thing. To make
+that work, Alice splits up her attachments into chunks and signs and encrypts
+them individually, so that Bob can verify just the chunks that he needs. This
+is all done carefully to make sure no one can move the chunks around after the
+fact.
+
+Another issue that comes up with large files is that it's helpful to host them
+on third-party CDN's. One worry about CDN's though, even with encrypted files,
+is that we might not be able to delete files reliably when we want to. To help
+with that, Alice encrypts attachments with a new set of one-time-use keys, and
+she includes those keys in the attachment message body. That way deleting the
+attachment message body has the same effect as deleting the large file.
+
+## Nitty Gritty Details
+
+### Algorithms
+
+Message encryption is done with NaCl's
+[`crypto_secretbox`](https://nacl.cr.yp.to/secretbox.html) (XSalsa20,
+Poly1305), and signing is done with NaCl's
+[`crypto_sign`](https://nacl.cr.yp.to/sign.html) (Ed25519, SHA512). The order
+of operations is:
```

Comment:
"The sender (or deleter or editor) of a message performs the following steps to encrypt:"